packages/core/core/src/requests/AssetGraphRequest.js

```diff
@@ -220,14 +220,22 @@ export class AssetGraphBuilder {
 
   shouldVisitChild(node: AssetGraphNode, child: AssetGraphNode): boolean {
     if (this.shouldBuildLazily) {
-      if (node.type === 'asset') {
+      if (node.type === 'asset' && child.type === 'dependency') {
         node.requested =
           node.requested || this.requestedAssetIds.has(node.value.id);
+        let previouslyDeferred = node.deferred;
         node.deferred =
           !node.requested &&
           this.assetGraph
             .getIncomingDependencies(node.value)
             .every(dep => dep.isEntry || dep.isAsync);
+
+        if (!previouslyDeferred && node.deferred) {
+          this.assetGraph.markParentsWithHasDeferred(node);
+        } else if (previouslyDeferred && !node.deferred) {
+          this.assetGraph.unmarkParentsWithHasDeferred(node);
+        }
```

Comment:
@mischnic does this look right to you? It ends up like this, and when the second dependency is added, deferred and hasDeferred need to be removed:

<img width="606" alt="image" src="https://user-images.githubusercontent.com/19409/108604280-f6dfd280-737a-11eb-9b6b-751d6836ddc5.png">

Kinda weird that a node has both `deferred` and `hasDeferred` maybe? I thought about putting `deferred` on the dependency node instead of the asset node, but then we'd need to check `hasDeferred` instead of `deferred` when creating a bundle so we know it's a placeholder. But there are many ways an asset node could have `hasDeferred` (e.g. unused side effect free dependencies), and only one way it could have `deferred`.

Reply:
Does't that exact logic exist here already? https://github.com/parcel-bundler/parcel/blob/72ce767797d6b43a0f96716df8c29707f775e25d/packages/core/core/src/AssetGraph.js#L271-L275

> Kinda weird that a node has both deferred and hasDeferred maybe? 

So before your changes, this was the idea:

Every node needs to have `hasDeferred` so that we know whether to recheck (grand)children  for un-deferring. Then additionally, if `AssetGroup.deferred` is true, that asset group has no child (and the AssetGraphBuilder doesn't try to resolve the asset). `Dependency.deferred` is actually unused, Flow doesn't complain if I remove that (in v2).

Reply:
Yeah I copied the logic from there. In this case, however, it's the asset that's deferred rather than the asset group. We have to transform the first asset past an async dependency, and can stop at that point (no need to resolve dependencies unless the asset becomes un-deferred).

Reply:
I see.

But the asset isn't deferred, because you actually did transform it? It's rather the dependencies of that first asset that are deferred (= have no children, the resolver wasn't run). So it should rather be `asset: hasDeferred`, `dependency: deferred`

Reply:
Yeah I could do that. The issue I thought of was how to know if a bundle is a placeholder or not. Currently it uses `entryAssetNode.deferred` to determine that. We could look at `entryAssetNode.hasDeferred` instead, but I was worried that might also be true in other cases, like if the asset had some dependencies that had no used symbols. Do you think that's possible? Should we add another flag to the asset for that? Or maybe check if _all_ of the asset's dependencies are deferred or something?

Reply:
> but I was worried that might also be true in other cases, like if the asset had some dependencies that had no used symbols.

That can indeed happen:

<img width="932" alt="Bildschirmfoto 2021-02-22 um 12 09 14" src="https://user-images.githubusercontent.com/4586894/108700372-c88efe00-7506-11eb-814c-ab47bee27c88.png">


> Or maybe check if all of the asset's dependencies are deferred or something?

I think the most consistent solution would be that (because then it would still hold that `deferred` = a leaf node). 

If that doesn't work out, a different flag might be better. (This reminds me of the `'references'` edge situation)

Reply:
k updated. Ended up removing `deferred` from asset nodes, and just using the `requested` property I already added. Made it so explicit `false` means it was deferred, `true` means it was previously deferred and then requested, and `undefined` means neither.